js/utils/serverConnect.js

```diff
@@ -0,0 +1,339 @@
+}
+
+export function getDebugLog() {
+  return debugLog;
+}
+
+export function getCurrentConnection() {
+  return currentConnection;
+}
+
+function socketConnect(socket) {
+  // todo: validate args
+
+  const deferred = $.Deferred();
+
+  const onOpen = () => deferred.resolve();
+  const onClose = (e) => deferred.reject('failed', e);
+
+  const cancel = () => {
+    socket.off(null, onOpen);
+    socket.off(null, onClose);
+    deferred.reject('canceled');
+  };
+
+  socket.on('open', onOpen);
+  socket.on('close', onClose);
+  socket.connect();
+
+  const promise = deferred.promise();
+  promise.cancel = cancel;
+
+  return promise;
+}
+
+function authenticate(server) {
+  // todo: validate args
+
+  const deferred = $.Deferred();
+
+  const fetchConfig = $.get(`${server.httpUrl}/ob/config`)
+    .done(() => deferred.resolve())
+    .fail((e) => deferred.reject('failed', e));
+
+  const cancel = () => {
+    deferred.reject('canceled');
+    fetchConfig.abort();
+  };
+
+  const promise = deferred.promise();
+  promise.cancel = cancel;
+
+  return promise;
+}
+
+export function connect(server, options = {}) {
+  if (!server instanceof ServerConfig) {
+    throw new Error('Please provide a server as a ServerConfig instance.');
+  }
+
+  // const curCon = getCurrentConnection();
+
+  // if (curCon && curCon.server.id === server.id) {
+  //   throw new Error('You are already connected to the given server');
+  // }
+
+  log(`[${server.id.slice(0, 8)}] Will attempt to connect to server "${server.get('name')}"` +
+    ` at ${server.get('serverIp')}.`);
+
+  const opts = {
+    attempts: 5,
+    timeoutBetweenAttempts: 2000,
+    maxAttemptTime: 5000,
+    // todo: work this one in
+    restartDefaultServerOnFirstAttempt: true,
+    ...options,
+  };
+
+  const deferred = $.Deferred();
+  const localServer = getLocalServer();
+  let attempt = 1;
+  let socket = null;
+  let connectAttempt = null;
+
+  const getPromiseData = (status, data = {}) => {
+    const aggregateData = {
+      status,
+      localServer,
+      server,
+      socket,
+      connectAttempt: attempt,
+      totalConnectAttempts: opts.attempts,
+      ...data,
+    };
+
+    if (localServer) aggregateData.localServer = localServer;
+
+    return aggregateData;
+  };
+
+  const notify = (e, ...args) => {
+    events.trigger(e.status, getPromiseData(e, ...args));
+    deferred.notify(getPromiseData(e, ...args));
+  };
+
+  const reject = (e, ...args) => {
+    events.trigger('connect-attempt-failed', getPromiseData(e, ...args));
+    deferred.reject(getPromiseData(e, ...args));
+  };
+
+  const resolve = (e, ...args) => {
+    events.trigger('connected', getPromiseData(e, ...args));
+    deferred.resolve(getPromiseData(e, ...args));
+  };
+
+  const cancel = () => {
+    if (socket) {
+      socket.off();
+      socket.close();
+    }
+
+    if (connectAttempt) connectAttempt.cancel();
+    reject('canceled');
+  };
+
+  // if (curCon) {
```

Comment:
I'll clean up stuff like this in the subsequent PR. Some of the commented code is stuff I'll actually revive and use.